File: backend/app/models/approval.py
from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field

from backend.app.models.action import RecommendedAction
from backend.app.models.context import SAPContext
from backend.app.models.diagnosis import DiagnosisResult
from backend.app.models.simulation import SimulationResult


class ApprovalStatus(str, Enum):
    PROPOSED = "proposed"
    AWAITING_APPROVAL = "awaiting_approval"
    APPROVED = "approved"
    REJECTED = "rejected"
    EXECUTED = "executed"
    ROLLED_BACK = "rolled_back"
    EXPIRED = "expired"


class ExecutionResult(BaseModel):
    success: bool
    sap_document_number: str | None = None
    message: str = ""
    executed_at: datetime = Field(default_factory=datetime.utcnow)
    execution_log: list[str] = Field(default_factory=list)


class ApprovalRequest(BaseModel):
    request_id: str
    context: SAPContext
    diagnosis: DiagnosisResult
    recommended_action: RecommendedAction
    simulation: SimulationResult
    status: ApprovalStatus = ApprovalStatus.PROPOSED
    requested_by: str | None = None
    requested_at: datetime = Field(default_factory=datetime.utcnow)
    approver: str | None = None
    approval_timestamp: datetime | None = None
    rejection_reason: str | None = None
    execution_result: ExecutionResult | None = None
    expires_at: datetime | None = None

File: frontend/src/types/approval.ts
import type { SAPContext, DiagnosisResult, RecommendedAction } from './sap'
import type { SimulationResult } from './simulation'

export type ApprovalStatus =
  | 'proposed'
  | 'awaiting_approval'
  | 'approved'
  | 'rejected'
  | 'executed'
  | 'rolled_back'
  | 'expired'

export interface ExecutionResult {
  success: boolean
  sap_document_number: string | null
  message: string
  executed_at: string
  execution_log: string[]
}

export interface ApprovalRequest {
  request_id: string
  context: SAPContext
  diagnosis: DiagnosisResult
  recommended_action: RecommendedAction
  simulation: SimulationResult
  status: ApprovalStatus
  requested_by: string | null
  requested_at: string
  approver: string | null
  approval_timestamp: string | null
  rejection_reason: string | null
  execution_result: ExecutionResult | null
  expires_at: string | null
}

export interface ApprovalResponse {
  request_id: string
  status: ApprovalStatus
  message: string
}

export interface ApprovalListResponse {
  total: number
  items: ApprovalRequest[]
}

export interface ApprovePayload {
  request_id: string
  approver: string
  comment?: string
}

export interface RejectPayload {
  request_id: string
  approver: string
  reason: string
}

export interface ExecutePayload {
  request_id: string
  executor: string
}

export const STATUS_LABEL: Record<ApprovalStatus, string> = {
  proposed: 'Proposed',
  awaiting_approval: 'Awaiting Approval',
  approved: 'Approved',
  rejected: 'Rejected',
  executed: 'Executed',
  rolled_back: 'Rolled Back',
  expired: 'Expired',
}

export const WORKFLOW_STEPS: ApprovalStatus[] = [
  'proposed',
  'awaiting_approval',
  'approved',
  'executed',
]
